File: 113_Path_Sum_II/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(root){
            solu.push_back(root->val);
            search(root,sum);
        }
        return res;
    }
    void search(TreeNode *root, int sum){
        if(!root->left&&!root->right)
            if(root->val == sum)
                    res.push_back(solu);
        if(root->left){
            solu.push_back(root->left->val);
            search(root->left,sum-root->val);
            solu.pop_back();
        }
        if(root->right){
            solu.push_back(root->right->val);
            search(root->right,sum-root->val);
            solu.pop_back();
        }
    }
private:
    vector<vector<int>> res;
    vector<int> solu;
};

File: 113_Path_Sum_II/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> solu = new ArrayList<Integer>();
        dfs(root, res, solu, sum, 0);
        return res;
    }
    
    private void dfs(TreeNode root, List<List<Integer>> res, List<Integer> solu, int sum, int current){
        if(root == null ) return ;
        current = current + root.val;
        
        solu.add(root.val);
        if(root.left == null && root.right == null && current == sum){
            res.add(new ArrayList<Integer>(solu));
            solu.remove(solu.size()-1);
            return ;
        }
        dfs(root.left, res, solu, sum, current);
        dfs(root.right, res, solu, sum, current);
        
        solu.remove(solu.size()-1);
    }
}